near-sdk-macros/src/lib.rs

```diff
@@ -213,6 +213,140 @@ pub fn metadata(item: TokenStream) -> TokenStream {
     }
 }
 
+#[proc_macro_derive(NearAbi, attributes(abi, serde, borsh_skip, schemars, validate))]
+pub fn derive_near_abi(_input: TokenStream) -> TokenStream {
+    // todo! borsh & serde don't support unions
+    // todo! guess what? we don't either
+    #[cfg(not(feature = "abi"))]
+    return TokenStream::new();
+    #[cfg(feature = "abi")]
+    {
```

Comment:
This functionality seems like it shouldn't be behind a feature, the attribute should just be gated behind the feature flag. Definitely a foot gun that a developer adds this and it does a no-op if they don't realize they didn't add the `abi` feature?

Or is the reason for this that there will be a plan to add the `abi` feature to the dependency automatically in `cargo-near`?

Either way, it seems like if the feature isn't set, it should maybe still implement `BorshSchema` and `Schemafy`, just not expose the `__near_abi` functions.

I don't have a high confidence in any of these suggestions, @itegulov definitely has the most context on this

Reply:
> Or is the reason for this that there will be a plan to add the abi feature to the dependency automatically in cargo-near?

Yeah, so I mean, unless there's a reason I'm missing for contracts having to manually specify the `abi` feature. I think it's only useful in the context of `cargo-near`, so we can have that automatically activate the feature.

> Definitely a foot gun that a developer adds this and it does a no-op if they don't realize they didn't add the abi feature?

If the developer's intention for deriving adding the trait is to generate abi, they have to use `cargo-near` and with the current state of `cargo-near`, it already won't let you do anything without the `abi` feature.

But rethinking this in line with your other commend, we should just rename the trait and leave it un-feature-gated.

Reply:
> Yeah, so I mean, unless there's a reason I'm missing for contracts having to manually specify the `abi` feature. I think it's only useful in the context of `cargo-near`, so we can have that automatically activate the feature.

Note that this is not currently how `cargo-near` does it

> If the developer's intention for deriving the trait is to generate abi, they have to use `cargo-near,` and with the current state of `cargo-near`, it already won't let you do anything without the `abi` feature.
> 
> But rethinking this in line with your other comment, we should just rename the trait and leave it un-feature-gated.

The only thing here is that it would be helpful to show errors for why it can't be derived without using the `abi` feature. For example, if a sub-type that's used in the struct does not implement the relevant trait, there would be no error until someone ran a command through `cargo-near`, which would be harder to fix.

Reply:
> Note that this is not currently how cargo-near does it

Yeah, I'm saying it's how it _could_ do it.

> there would be no error until someone ran a command through cargo-near, which would be harder to fix.

Good point.

Reply:
So we're to get to the point where;

1. contracts don't need to manually specify the "abi" feature flag
2. `DeriveName` is not feature gated behind the "abi" feature flag, but its implementors are only injected when the target isn't wasm.

This would ensure `DeriveName` can be debugged without `cargo-near`.

Reply:
My idea for "abi" feature flag is that it will be gone once we stabilize the ABI schema and devx around it in general. So yeah @miraclx your last comment describes how I see it perfectly.

> but its implementors are only injected when the target isn't wasm.

Side note, but we could even create a custom rustc target for ABI so that it does not clash with unit tests. This would also provide a uniform experience on all user platforms.

Reply:
> This would be difficult to achieve btw, as it will require either modifying the user's Cargo.toml file

Not really, `cargo build --features near-sdk/abi` works to the same effect as `near-sdk = { ..., features = ["abi"]}`

Although, I'm taking a new approach now. If you take a look at https://github.com/near/cargo-near/pull/31, what I resolved to do instead is:

- Contracts manually specify the `abi` feature flag if they want to generate ABI.
   - This gives them access to the `NearSchema` trait that they can derive on their structures. 
- `cargo-near` rejects all contracts that don't have the `abi` feature flag.
- `cargo-near` activates the `__abi-generate` feature flag when generating ABI.
- `cargo-near` activates the `__abi-embed` feature flag when embedding ABI.

(leading underscores is to ensure docs.rs hides those features, and leaving the hyphen inconsistency in there to signal contract devs not to use it)

Reply:
> Not really, cargo build --features near-sdk/abi works to the same effect as near-sdk = { ..., features = ["abi"]}

lol, how did I not think of that

> Although, I'm taking a new approach now. If you take a look at https://github.com/near/cargo-near/pull/31

Interesting, I will definitely take a look